plugin/pkg/admission/webhook/config.go

```
@@ -0,0 +1,126 @@
+	errRetryBackoffOutOfRange = "webhook retry backoff is invalid: %v should be between %v and %v"
+	minRetryBackoff           = time.Duration(1) * time.Millisecond
+	maxRetryBackoff           = time.Duration(5) * time.Minute
+)
+
+// DefaultRule is a rule that will skip processing by default
+var (
+	DefaultRule = Rule{Type: Skip}
+)
+
+func normalizeConfig(config *GenericAdmissionWebhookConfig) error {
+	allowFailAction := string(Allow)
+	denyFailAction := string(Deny)
+	connectOperation := string(admission.Connect)
+	createOperation := string(admission.Create)
+	deleteOperation := string(admission.Delete)
+	updateOperation := string(admission.Update)
+	sendRuleType := string(Send)
+	skipRuleType := string(Skip)
+
+	// Validate the kubeConfigFile property is present
+	if config.KubeConfigFile == "" {
+		return fmt.Errorf(errMissingKubeConfigFile)
+	}
+
+	// Normalize and validate the retry backoff
+	if config.RetryBackoff == defaultDuration {
+		config.RetryBackoff = defaultRetryBackoff
+	} else {
+		// Unmarshalling gives nanoseconds so convert to milliseconds
+		config.RetryBackoff *= time.Millisecond
+	}
+
+	if config.RetryBackoff < minRetryBackoff || config.RetryBackoff > maxRetryBackoff {
+		return fmt.Errorf(errRetryBackoffOutOfRange, config.RetryBackoff, minRetryBackoff, maxRetryBackoff)
+	}
+
+	// Validate that there is at least one rule
+	if len(config.Rules) == 0 {
+		return fmt.Errorf(errOneRuleRequired)
+	}
+
+	for i, rule := range config.Rules {
+		// Normalize and validate the fail action
+		failAction := strings.ToUpper(string(rule.FailAction))
+
+		switch failAction {
+		case "":
+			config.Rules[i].FailAction = defaultFailAction
+		case allowFailAction:
+			config.Rules[i].FailAction = Allow
+		case denyFailAction:
+			config.Rules[i].FailAction = Deny
+		default:
+			return fmt.Errorf(errInvalidFailAction, i, rule.FailAction, Allow, Deny)
+		}
+
+		// Normalize and validate the rule operation(s)
```

We usually validate and fail fast, rather than allow incorrect data.